python/pyspark/sql/dataframe.py

```diff
@@ -305,11 +305,11 @@ def explain(self, extended=None, mode=None):
         is_mode_case = mode is not None and isinstance(mode, basestring)
 
         if not is_no_argument and not (is_extended_case or is_mode_case):
-            argtypes = [
-                str(type(arg)) for arg in [extended, mode] if arg is not None]
-            raise TypeError(
-                "extended (optional) and mode (optional) should be a bool and str; "
-                "however, got [%s]." % ", ".join(argtypes))
+            if extended is not None:
+                errMsg = "extended should be provided as bool, got {0}".format(type(extended))
+            else:  # For mode case
+                errMsg = "mode should be provided as str, got {0}".format(type(mode))
+            raise TypeError(errMsg)
```

Comment:
How about this error handling? @viirya https://github.com/apache/spark/pull/26861#discussion_r357726827

Reply:
either `extended (optional) should be ...` or `extended, if given, should be ...`. same for `mode`.

Reply:
Ah, ok. I'll update. (but, I'm away from my pc, so do tonight.)

Reply:
Does the latest commit satisfy you comment? The current exception behaviours are as follows;
```
>>> spark.range(1).explain(extended=1)
TypeError: extended (optional) should be provided as bool, got <class 'int'>

>>> spark.range(1).explain(mode=1)
TypeError: mode (optional) should be provided as str, got <class 'int'>

>>> spark.range(1).explain(extended=1, mode=1)
Exception: extended and mode can not be specified simultaneously
```

Reply:
LGTM.